C compilation error with option array in embedded struct

### Describe the bug

Use of an option typed array in an embedded struct can lead to C compiler error.

### Reproduction Steps

```v
pub struct Struct1 {
pub mut:
	strings ?[]string
}

@[params]
pub struct Struct2 {
	Struct1
}

fn my_func(args Struct2) {
	mut s1 := args.Struct1
	s1.strings << ['hi']
}

s2 := Struct2{}

my_func(s2)
```

### Expected Behavior

No C compiler error.

### Current Behavior

```console
$ v run array.v
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/array.01JQEYGMFER25W82AS45XKVG2E.tmp.c:6010: error: '{' expected (got ";")
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

0.4.10

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.10 a59ebea.fcd2eed
|:-------------------|:-------------------
|OS                  |linux, "Arch Linux"
|Processor           |8 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8350U CPU @ 1.70GHz
|Memory              |0.4GB/23.22GB
|                    |
|V executable        |/home/scott/builds/v/v
|V last modified time|2025-03-28 17:58:59
|                    |
|V home dir          |OK, value: /home/scott/builds/v
|VMODULES            |OK, value: /home/scott/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/scott/threefold/repos/herolib
|                    |
|Git version         |git version 2.48.1
|V git status        |weekly.2023.44-3276-gfcd2eed0
|.git/config present |true
|                    |
|cc version          |cc (GCC) 14.2.1 20240910
|gcc version         |gcc (GCC) 14.2.1 20240910
|clang version       |clang version 19.1.7
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.40

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -118,7 +118,7 @@ mut:
 	chan_pop_options          map[string]string // types for `x := <-ch or {...}`
 	chan_push_options         map[string]string // types for `ch <- x or {...}`
 	mtxs                      string            // array of mutexes if the `lock` has multiple variables
-	tmp_var_is_ptr            bool              // indicates if the tmp var passed to or_block() is a ptr
+	tmp_var_ptr               map[string]bool   // indicates if the tmp var passed to or_block() is a ptr
 	labeled_loops             map[string]&ast.Stmt
 	contains_ptr_cache        map[ast.Type]bool
 	inner_loop                &ast.Stmt = unsafe { nil }
@@ -4118,10 +4118,13 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		if needs_addr {
 			g.write(')')
 		}
-
-		g.tmp_var_is_ptr = is_option_unwrap
+		if is_option_unwrap {
+			g.tmp_var_ptr[tmp_var] = true
+		}
 		g.or_block(tmp_var, node.or_block, node.typ)
-		g.tmp_var_is_ptr = false
+		if is_option_unwrap {
+			g.tmp_var_ptr.delete(tmp_var)
+		}
 		g.write2(stmt_str, ' ')
 		unwrapped_typ := node.typ.clear_option_and_result()
 		unwrapped_styp := g.styp(unwrapped_typ)
@@ -6944,7 +6947,7 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 					mut is_array_fixed := false
 					mut return_wrapped := false
 					mut return_is_option := is_option && return_type.has_option_or_result()
-					tmp_op := if g.tmp_var_is_ptr { '->' } else { '.' }
+					tmp_op := if cvar_name in g.tmp_var_ptr { '->' } else { '.' }
 					if is_option {
 						is_array_fixed = g.table.final_sym(return_type).kind == .array_fixed
 						if !is_array_fixed {
@@ -7008,7 +7011,7 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 // Returns the type of the last stmt
 fn (mut g Gen) or_block(var_name string, or_block ast.OrExpr, return_type ast.Type) {
 	cvar_name := c_name(var_name)
-	tmp_op := if g.tmp_var_is_ptr { '->' } else { '.' }
+	tmp_op := if var_name in g.tmp_var_ptr { '->' } else { '.' }
 	if or_block.kind == .block && or_block.stmts.len == 0 {
 		// generate nothing, block is empty
 		g.write(';\n${util.tabs(g.indent)}(void)${cvar_name};')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4082,8 +4082,13 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		if is_ptr {
 			g.write('*(')
 		}
+		needs_addr := is_option_unwrap && node.expr !in [ast.Ident, ast.PrefixExpr]
 		if is_option_unwrap {
-			g.write('&')
+			if !needs_addr {
+				g.write('&')
+			} else {
+				g.write('ADDR(${styp}, ')
+			}
 		}
 		g.expr(node.expr)
 		for i, embed in node.from_embed_types {
@@ -4108,6 +4113,9 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		}
 		g.write(field_name)
 		if is_ptr {
+			g.write(')')
+		}
+		if needs_addr {
 			g.write(')')
 		}
 

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6948,7 +6948,7 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 								return_wrapped = true
 							} else if expr_stmt.expr is ast.CallExpr {
 								if expr_stmt.expr.is_return_used {
-									g.write('*(${cast_typ}*) ${cvar_name}.data = ')
+									g.write('*(${cast_typ}*) ${cvar_name}${tmp_op}data = ')
 								}
 							} else if g.inside_opt_or_res && return_is_option && g.inside_assign {
 								g.write('_option_ok(&(${cast_typ}[]) { ')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6936,6 +6936,7 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 					mut is_array_fixed := false
 					mut return_wrapped := false
 					mut return_is_option := is_option && return_type.has_option_or_result()
+					tmp_op := if g.tmp_var_is_ptr { '->' } else { '.' }
 					if is_option {
 						is_array_fixed = g.table.final_sym(return_type).kind == .array_fixed
 						if !is_array_fixed {
@@ -6956,14 +6957,14 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 								g.indent--
 								return
 							} else {
-								g.write('*(${cast_typ}*) ${cvar_name}.data = ')
+								g.write('*(${cast_typ}*) ${cvar_name}${tmp_op}data = ')
 							}
 						}
 					} else {
 						g.write('${cvar_name} = ')
 					}
 					if is_array_fixed {
-						g.write('memcpy(${cvar_name}.data, (${cast_typ})')
+						g.write('memcpy(${cvar_name}${tmp_op}data, (${cast_typ})')
 					}
 					// return expr or { fn_returns_option() }
 					if is_option && g.inside_return && expr_stmt.expr is ast.CallExpr

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -118,6 +118,7 @@ mut:
 	chan_pop_options          map[string]string // types for `x := <-ch or {...}`
 	chan_push_options         map[string]string // types for `ch <- x or {...}`
 	mtxs                      string            // array of mutexes if the `lock` has multiple variables
+	tmp_var_is_ptr            bool              // indicates if the tmp var passed to or_block() is a ptr
 	labeled_loops             map[string]&ast.Stmt
 	contains_ptr_cache        map[ast.Type]bool
 	inner_loop                &ast.Stmt = unsafe { nil }
@@ -4071,10 +4072,18 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		stmt_str := g.go_before_last_stmt().trim_space()
 		styp := g.styp(g.unwrap_generic(node.typ))
 		g.empty_line = true
+		is_option_unwrap := node.typ.has_flag(.option)
 		tmp_var := g.new_tmp_var()
-		g.write('${styp} ${tmp_var} = ')
+		g.write('${styp} ')
+		if is_option_unwrap {
+			g.write('*')
+		}
+		g.write('${tmp_var} = ')
 		if is_ptr {
 			g.write('*(')
+		}
+		if is_option_unwrap {
+			g.write('&')
 		}
 		g.expr(node.expr)
 		for i, embed in node.from_embed_types {
@@ -4101,11 +4110,14 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		if is_ptr {
 			g.write(')')
 		}
+
+		g.tmp_var_is_ptr = is_option_unwrap
 		g.or_block(tmp_var, node.or_block, node.typ)
+		g.tmp_var_is_ptr = false
 		g.write2(stmt_str, ' ')
 		unwrapped_typ := node.typ.clear_option_and_result()
 		unwrapped_styp := g.styp(unwrapped_typ)
-		g.write('(*(${unwrapped_styp}*)${tmp_var}.data)')
+		g.write('(*(${unwrapped_styp}*)${tmp_var}->data)')
 		return
 	}
 
@@ -6987,6 +6999,7 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 // Returns the type of the last stmt
 fn (mut g Gen) or_block(var_name string, or_block ast.OrExpr, return_type ast.Type) {
 	cvar_name := c_name(var_name)
+	tmp_op := if g.tmp_var_is_ptr { '->' } else { '.' }
 	if or_block.kind == .block && or_block.stmts.len == 0 {
 		// generate nothing, block is empty
 		g.write(';\n${util.tabs(g.indent)}(void)${cvar_name};')
@@ -6996,20 +7009,20 @@ fn (mut g Gen) or_block(var_name string, or_block ast.OrExpr, return_type ast.Ty
 	is_none_ok := return_type == ast.ovoid_type
 	g.writeln(';')
 	if is_none_ok {
-		g.writeln('if (${cvar_name}.state != 0) {')
+		g.writeln('if (${cvar_name}${tmp_op}state != 0) {')
 	} else {
 		if return_type != 0 && g.table.sym(return_type).kind == .function {
 			mr_styp = 'voidptr'
 		}
 		if return_type.has_flag(.result) {
-			g.writeln('if (${cvar_name}.is_error) {')
+			g.writeln('if (${cvar_name}${tmp_op}is_error) {')
 		} else {
-			g.writeln('if (${cvar_name}.state != 0) {')
+			g.writeln('if (${cvar_name}${tmp_op}state != 0) {')
 		}
 	}
 	if or_block.kind == .block {
 		g.or_expr_return_type = return_type.clear_option_and_result()
-		g.writeln('\tIError err = ${cvar_name}.err;')
+		g.writeln('\tIError err = ${cvar_name}${tmp_op}err;')
 
 		g.inside_or_block = true
 		defer {
@@ -7029,7 +7042,7 @@ fn (mut g Gen) or_block(var_name string, or_block ast.OrExpr, return_type ast.Ty
 		|| (or_block.kind == .propagate_option && return_type.has_flag(.result)) {
 		if g.file.mod.name == 'main' && (g.fn_decl == unsafe { nil } || g.fn_decl.is_main) {
 			// In main(), an `opt()!` call is sugar for `opt() or { panic(err) }`
-			err_msg := 'IError_name_table[${cvar_name}.err._typ]._method_msg(${cvar_name}.err._object)'
+			err_msg := 'IError_name_table[${cvar_name}${tmp_op}err._typ]._method_msg(${cvar_name}${tmp_op}err._object)'
 			if g.pref.is_debug {
 				paline, pafile, pamod, pafn := g.panic_debug_info(or_block.pos)
 				g.writeln('panic_debug(${paline}, tos3("${pafile}"), tos3("${pamod}"), tos3("${pafn}"), ${err_msg});')
@@ -7057,7 +7070,7 @@ fn (mut g Gen) or_block(var_name string, or_block ast.OrExpr, return_type ast.Ty
 	} else if or_block.kind == .propagate_option {
 		if g.file.mod.name == 'main' && (g.fn_decl == unsafe { nil } || g.fn_decl.is_main) {
 			// In main(), an `opt()?` call is sugar for `opt() or { panic(err) }`
-			err_msg := 'IError_name_table[${cvar_name}.err._typ]._method_msg(${cvar_name}.err._object)'
+			err_msg := 'IError_name_table[${cvar_name}${tmp_op}err._typ]._method_msg(${cvar_name}${tmp_op}err._object)'
 			if g.pref.is_debug {
 				paline, pafile, pamod, pafn := g.panic_debug_info(or_block.pos)
 				g.writeln('panic_debug(${paline}, tos3("${pafile}"), tos3("${pamod}"), tos3("${pafn}"), ${err_msg}.len == 0 ? _SLIT("option not set ()") : ${err_msg});')

--- a/vlib/v/gen/c/cmain.v
+++ b/vlib/v/gen/c/cmain.v
@@ -249,7 +249,7 @@ pub fn (mut g Gen) gen_failing_error_propagation_for_test_fn(or_block ast.OrExpr
 	// `or { cb_propagate_test_error(@LINE, @FILE, @MOD, @FN, err.msg() ) }`
 	// and the test is considered failed
 	paline, pafile, pamod, pafn := g.panic_debug_info(or_block.pos)
-	dot_or_ptr := if g.tmp_var_is_ptr { '->' } else { '.' }
+	dot_or_ptr := if cvar_name in g.tmp_var_ptr { '->' } else { '.' }
 	err_msg := 'IError_name_table[${cvar_name}${dot_or_ptr}err._typ]._method_msg(${cvar_name}${dot_or_ptr}err._object)'
 	g.writeln('\tmain__TestRunner_name_table[test_runner._typ]._method_fn_error(test_runner._object, ${paline}, tos3("${pafile}"), tos3("${pamod}"), tos3("${pafn}"), ${err_msg} );')
 	g.writeln('\tlongjmp(g_jump_buffer, 1);')
@@ -260,7 +260,7 @@ pub fn (mut g Gen) gen_failing_return_error_for_test_fn(return_stmt ast.Return, 
 	// `or { err := error('something') cb_propagate_test_error(@LINE, @FILE, @MOD, @FN, err.msg() ) return err }`
 	// and the test is considered failed
 	paline, pafile, pamod, pafn := g.panic_debug_info(return_stmt.pos)
-	dot_or_ptr := if g.tmp_var_is_ptr { '->' } else { '.' }
+	dot_or_ptr := if cvar_name in g.tmp_var_ptr { '->' } else { '.' }
 	err_msg := 'IError_name_table[${cvar_name}${dot_or_ptr}err._typ]._method_msg(${cvar_name}${dot_or_ptr}err._object)'
 	g.writeln('\tmain__TestRunner_name_table[test_runner._typ]._method_fn_error(test_runner._object, ${paline}, tos3("${pafile}"), tos3("${pamod}"), tos3("${pafn}"), ${err_msg} );')
 	g.writeln('\tlongjmp(g_jump_buffer, 1);')

--- a/vlib/v/gen/c/cmain.v
+++ b/vlib/v/gen/c/cmain.v
@@ -249,7 +249,8 @@ pub fn (mut g Gen) gen_failing_error_propagation_for_test_fn(or_block ast.OrExpr
 	// `or { cb_propagate_test_error(@LINE, @FILE, @MOD, @FN, err.msg() ) }`
 	// and the test is considered failed
 	paline, pafile, pamod, pafn := g.panic_debug_info(or_block.pos)
-	err_msg := 'IError_name_table[${cvar_name}.err._typ]._method_msg(${cvar_name}.err._object)'
+	dot_or_ptr := if g.tmp_var_is_ptr { '->' } else { '.' }
+	err_msg := 'IError_name_table[${cvar_name}${dot_or_ptr}err._typ]._method_msg(${cvar_name}${dot_or_ptr}err._object)'
 	g.writeln('\tmain__TestRunner_name_table[test_runner._typ]._method_fn_error(test_runner._object, ${paline}, tos3("${pafile}"), tos3("${pamod}"), tos3("${pafn}"), ${err_msg} );')
 	g.writeln('\tlongjmp(g_jump_buffer, 1);')
 }
@@ -259,7 +260,8 @@ pub fn (mut g Gen) gen_failing_return_error_for_test_fn(return_stmt ast.Return, 
 	// `or { err := error('something') cb_propagate_test_error(@LINE, @FILE, @MOD, @FN, err.msg() ) return err }`
 	// and the test is considered failed
 	paline, pafile, pamod, pafn := g.panic_debug_info(return_stmt.pos)
-	err_msg := 'IError_name_table[${cvar_name}.err._typ]._method_msg(${cvar_name}.err._object)'
+	dot_or_ptr := if g.tmp_var_is_ptr { '->' } else { '.' }
+	err_msg := 'IError_name_table[${cvar_name}${dot_or_ptr}err._typ]._method_msg(${cvar_name}${dot_or_ptr}err._object)'
 	g.writeln('\tmain__TestRunner_name_table[test_runner._typ]._method_fn_error(test_runner._object, ${paline}, tos3("${pafile}"), tos3("${pamod}"), tos3("${pafn}"), ${err_msg} );')
 	g.writeln('\tlongjmp(g_jump_buffer, 1);')
 }

--- a/vlib/v/tests/options/option_array_push_test.v
+++ b/vlib/v/tests/options/option_array_push_test.v
@@ -1,0 +1,22 @@
+pub struct Struct1 {
+pub mut:
+	strings ?[]string
+}
+
+fn test_main() {
+	mut s1 := Struct1{}
+	s1.strings = []
+	s1.strings? << ['foo']
+	s1.strings? << ['bar']
+
+	dump(s1)
+	assert s1.strings? == ['foo', 'bar']
+
+	assert s1.strings? == ['foo', 'bar']
+
+	s1.strings = []
+	assert s1.strings? == []
+
+	s1.strings? << ['foo']
+	assert s1.strings? == ['foo']
+}

--- a/vlib/v/tests/options/option_array_push_test.v
+++ b/vlib/v/tests/options/option_array_push_test.v
@@ -5,13 +5,12 @@ pub mut:
 
 fn test_main() {
 	mut s1 := Struct1{}
+	assert s1.strings == none
 	s1.strings = []
+	assert s1.strings? == []
 	s1.strings? << ['foo']
 	s1.strings? << ['bar']
-
 	dump(s1)
-	assert s1.strings? == ['foo', 'bar']
-
 	assert s1.strings? == ['foo', 'bar']
 
 	s1.strings = []

